### DISCUSSION TITLE
How do I get a URL from a thumbnail?

### DISCUSSION THREAD

**Discussion by Veinhelm** [ClementJ18/moddb#24] [General]
Title says it all. Making something where a random Half-Life 1 mod is picked and gets the url.

**Comment by ClementJ18** [comment]
The thumbnail object has a `url` attribute you can use: https://moddb.readthedocs.io/en/stable/boxes.html#thumbnail

Feel free to post some code if anything is unclear

**Reply by Veinhelm** [reply]
I've tried as many things as I can think of, and I can only get the name of the mod. I have to do some stupid stuff to get the name. I'm not sure if I'm dumb, or if the library is underdeveloped.

Here's the code. It's for a Discord bot.
```python
if message.content.lower() == ">randmod":
	print(f"{message.author.name}#{message.author.discriminator} has called >randmod")
	bm = await message.channel.send("Checking ModDB...")
	hl = moddb.parse_page("https://www.moddb.com/games/half-life")
	mod = random.choice(moddb.Game.get_mods(hl, random.randint(1, 40)))
	mn1 = str.replace(str(mod), "<Thumbnail name=", "")
	mn2 = str.replace(mn1, " type=mod>", "")
	await bm.edit(content=mn2)
```

**Reply by ClementJ18** [reply]
See how you access the author of the discord message using `.author`? You can do something similar with the thumbnail objects returned by get_mods. Get mods returns a `list` of `moddb.Thumbnail`  objects which all have a `url`  attribute. 

First, you should cache the game object, parse the page once on start up and attach it to your bot instance. Secondly, when using get_mod you can call it using the instance of Game, so you can just do `hl.get_mods`. Once you have an instance of something you can use all the methods documented.

Finally, the methods like `parse_page` and `get_mods`  are blocking (See more [here](https://discordpy.readthedocs.io/en/stable/faq.html#what-does-blocking-mean)). In the future you may want to refactor the code similarly to how it is explained [here](https://moddb.readthedocs.io/en/stable/#advanced-usage). This would look something like:
```python
async with aiohttp.ClientSession() as session:
      async with session.get("https://www.moddb.com/games/half-life/mods/") as response:
         soup = moddb.soup(await r.text())

thumbnails, _, _, _ = moddb..boxes._parse_results(soup)
thumbnail = random.choice(thumbnails)
```

**Reply by Veinhelm** [reply]
Sorry for this late response. I procrastinated from doing this for a while and finally decided to do it now. Anyways, what you said to do worked. There is one problem though. It only selects mods on the first page. Is there any way of making it choose from any page? I searched "page" on the docs and nothing of use came up.

Here's what I have, if it helps.

```python
if message.content.lower() == ">randmod":
	print(f"{message.author.name}#{message.author.discriminator} has called >randmod")

	bm = await message.channel.send("Checking ModDB...")

	async with aiohttp.ClientSession() as session:
		async with session.get("https://www.moddb.com/games/half-life/mods") as response:
			soup = moddb.soup(await response.text())

	thumbnails, _, _, _ = moddb.boxes._parse_results(soup)
	thumbnail = random.choice(thumbnails)

	await bm.edit(content=thumbnail.url)
```

**Reply by ClementJ18** [reply]
The regular library has methods for pagination but since you're using the raw methods due to the async context you'll have to handle it yourself. `_parse_results` returns pagination data you can use. The 2nd element returned is the current page, the 3rd the max page number and the 4th the number of mods.

One way of doing it is iterating over every page until you've gathered every mod, however, that will get your IP banned very quickly. My recommendation is to instead calculate the number of mods using the total results and pick a random number between 0 and the max amount. You can then go directly to the page of that mod by diving by 30 and you can figure out which mod was picked using modulo.


picked_number // 30 = page number
picked_number % 30 = mod index on that page

This is an interesting discussion, it's giving me a few ideas to support people like you who can't use the regular sync methods.

**Reply by Veinhelm** [reply]
I'm back again, lmao. I made my own way of checking every page and I thought why not share it.
Basically, I thought that maybe I could just get the number of pages from the navbar at the bottom of the page, the one that is like `1 2 3 4 5 6 7 8 ... (last page)`
I put together this function which uses Beautiful Soup to find the "...", which is in a span with the class "spacer", and then get the next element's text and convert it to an integer, which would be the last page.
Then, in the bot command, the url it chooses to select the mod from is a formatted string that picks the page from 1 to whatever the last page number is.

Here's the code:
```python
async def getHalfLifeModPages():
	async with aiohttp.ClientSession() as s:
		async with s.get("https://www.moddb.com/games/half-life/mods") as r:
			soup = BeautifulSoup(await r.text(), features="html.parser")
			spacer = soup.find("span", class_="spacer")
			lastpage = spacer.find_next("a").text
	return int(lastpage)
...
@tree.command(name="randmod", description="Selects and sends a random Half-Life 1 mod from ModDB")
async def randmod(interaction):
	print(f"{interaction.user} has called /{interaction.command.name}")
	await interaction.response.send_message("Checking ModDB...")
	async with aiohttp.ClientSession() as session:
		async with session.get(f"https://www.moddb.com/games/half-life/mods/page/{random.randint(1, await getHalfLifeModPages())}") as response:
			soup = moddb.soup(await response.text())
	
	thumbnails, _, _, _ = moddb.boxes._parse_results(soup)
	thumbnail = random.choice(thumbnails)

	await interaction.edit_original_response(content=thumbnail.url)
```

Maybe you could use this method or something similar to implement an "all pages" search...

Also, I'd like to add, for a while before this I was using the code you sent in your last comment, but it would pick a random integer from 1 to whatever the last page at the time was, meaning I would have to manually update it lol.

**Reply by ClementJ18** [reply]
If you want to get the page data from a list of results you have `get_list_stats`
```py
async def getHalfLifeModPages():
	async with aiohttp.ClientSession() as s:
		async with s.get("https://www.moddb.com/games/half-life/mods") as r:
			soup = moddb.soup(await r.text())

        page_stats = get_list_stats(soup)
	return page_stat[1]
```

`_parse_results` also returns pagination data. 

I don't get your last sentence. What did you want the code to do?
